File: Stack/20.py
"""
20. 有效的括号
简单
https://leetcode-cn.com/problems/valid-parentheses/
"""

class Solution:
    # 方法一: 利用栈，遇到左括号入栈，遇到右括号将栈中的栈顶元素出栈，并判断当前的括号和出栈的括号是否匹配。
    # 40 ms	    75.79%
    # 13.9 MB	5.22%
    def isValid(self, s: str) -> bool:
        if not s:
            return True

        parenthesisDict = {
            '(': ')',
            '[': ']',
            '{': '}'
        }
        keys = parenthesisDict.keys()
        stack = []

        for i in s:
            if i in keys:
                stack.append(i)
            else:
                if len(stack) == 0:
                    return False
                top = stack.pop()
                if i != parenthesisDict[top]:
                    return False
        
        return len(stack) == 0

File: Leetcode-Swift/Leetcode/20.swift
//
//  20.valid-parentheses.swift
//  Leetcode-Swift
//
//  Created by devshiye on 2021/4/4.
//

import Foundation
/*
 20. 有效的括号 (简单)
 https://leetcode-cn.com/problems/valid-parentheses/
 */

class Solution_20 {
    func isValid(_ s: String) -> Bool {
        if s.isEmpty {
            return true
        }
        let dict: [Character: Character] = [
            "(": ")",
            "[": "]",
            "{": "}"
            ]
        var stack = [Character]()
        for ch in s {
            if let _ = dict[ch] {
                // 如果是左括号入栈
                stack.append(ch)
            } else {
                // 如果是右括号，应该与栈顶的左括号进行匹配
                // 栈顶为空，直接返回false
                if stack.isEmpty {
                    return false
                }
                // 取出栈顶元素，判断是否和栈顶的左括号匹配
                let top = stack.removeLast()
                if let val = dict[top], val != ch {
                    return false
                }
            }
        }

        return stack.isEmpty
    }
    
    func isValid_old(_ s: String) -> Bool {
        var stack: [Character] = []
        let dict: [Character : Character] = ["{":"}", "(":")", "[":"]"]
        for char in s {
            if dict[char] != nil {
                stack.append(char)
            } else {
                if dict[stack.last ?? "x"] == char {
                    stack.removeLast()
                } else {
                    return false
                }
            }
        }
        return stack.count == 0
    }

    /*
     测试用例：
     1. 1个(左/右)、多个匹配，多个不匹配（奇数个数不匹配、偶数个数不匹配）
     2. 空
     */
    func test() {
        print(isValid("{"))
        print(isValid("]"))
        print(isValid("{]"))
        print(isValid("{}"))
        print(isValid("[]"))
        print(isValid("()"))
        print(isValid("{[()]}"))
        print(isValid("{{{}}}"))
        print(isValid("{{{[}}}"))
        print(isValid(""))
    }
}
